// @cubejs-backend/testing/dist/src/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCubeSchema = exports.getLocalHostnameByOs = void 0;
const os_1 = require("os");
function getLocalHostnameByOs() {
    if (os_1.platform() === 'win32') {
        return 'docker.for.win.localhost';
    }
    if (os_1.platform() === 'darwin') {
        return 'host.docker.internal';
    }
    return 'localhost';
}
exports.getLocalHostnameByOs = getLocalHostnameByOs;
function createCubeSchema({ name, refreshKey = '', preAggregations = '' }) {
    return ` 
    cube('${name}', {
        sql: \`
        select * from cards
        \`,
        
        ${refreshKey}
   
        measures: {
          count: {
            type: 'count'
          },
          sum: {
            sql: \`amount\`,
            type: \`sum\`
          },
          max: {
            sql: \`amount\`,
            type: \`max\`
          },
          min: {
            sql: \`amount\`,
            type: \`min\`
          }
        },
  
        dimensions: {
          id: {
            type: 'number',
            sql: 'id',
            primaryKey: true
          },
          createdAt: {
            type: 'time',
            sql: 'created_at'
          },
        },
        
        preAggregations: {
            ${preAggregations}
        }
      }) 
  `;
}
exports.createCubeSchema = createCubeSchema;
